### ISSUE TITLE
Redirects to google login after successful authentication

### ISSUE THREAD

**Issue by EmiPhil** [thomseddon/traefik-forward-auth#92]
Hi, I'm struggling to get this configured properly. Using traefik v2 in k8s. I navigate to app.dev.host.solutions and correctly get redirected to the google login, but then get redirected back to the google login again instead of app.dev.host.solutions.

O-auth config:

```yaml
kind: Service
apiVersion: v1
metadata:
  name: traefik-forward-auth-developers
spec:
  selector:
    app: traefik-forward-auth-developers
  ports:
    - protocol: TCP
      port: 4181
  type: ClusterIP
---
kind: Deployment
apiVersion: apps/v1
metadata:
  namespace: default
  name: traefik-forward-auth-developers
  labels:
    app: traefik-forward-auth-developers
spec:
  replicas: 1
  selector:
    matchLabels:
      app: traefik-forward-auth-developers
  template:
    metadata:
      labels:
        app: traefik-forward-auth-developers
        name: traefik-forward-auth-developers
    spec:
      containers:
        - name: traefik-forward-auth-developers
          image: thomseddon/traefik-forward-auth:2
          env:
            - name: PROVIDERS_GOOGLE_CLIENT_ID
              valueFrom:
                secretKeyRef:
                  name: g-dev-oauth
                  key: client-id
            - name: PROVIDERS_GOOGLE_CLIENT_SECRET
              valueFrom:
                secretKeyRef:
                  name: g-dev-oauth
                  key: client-secret
            - name: SECRET
              valueFrom:
                secretKeyRef:
                  name: g-dev-oauth
                  key: o-secret
            - name: AUTH_HOST
              value: auth.dev.host.solutions
            - name: COOKIE_DOMAIN
              value: dev.host.solutions
            - name: INSECURE_COOKIE
              value: "false"
            - name: LOG_LEVEL
              value: debug
          ports:
            - containerPort: 4181
              protocol: TCP
          resources:
            limits:
              cpu: "0.2"
              memory: 200Mi
          livenessProbe:
            tcpSocket:
              port: 4181
            initialDelaySeconds: 20
            failureThreshold: 3
            successThreshold: 1
            periodSeconds: 10
            timeoutSeconds: 2
```

Middleware:

```yaml
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: dev-auth
spec:
  forwardAuth:
    address: http://traefik-forward-auth-developers:4181/_oauth
    authResponseHeaders:
      - X-Forwarded-User
    trustForwardHeader: true
```

Route:

```yaml
apiVersion: traefik.containo.us/v1alpha1
kind: IngressRoute
metadata:
  name: dev-auth
  namespace: default
spec:
  entryPoints:
    - https
  routes:
    - match: Host(`auth.dev.host.solutions`)
      kind: Rule
      services:
        - name: traefik-forward-auth-developers
          port: 4181
  tls:
    secretName: dev-cert
```

At first I thought this may be similar to the issue described in #79 but I think it may be different. I tried changing the auth host, cookie domain and app host name to all be directly off host.solutions instead of off of dev.host.solutions but it didn't make a difference. I also built and deployed #81 but that also did not work.

Any ideas?

**Comment by qlaffont** [comment]
Same for me. I have check log and I have this error : "Error validating csrf cookie: Invalid CSRF state value"

**Comment by javaadpatel** [comment]
@EmiPhil , I was experiencing the login loop as well, this issue gave me a hint:
https://github.com/thomseddon/traefik-forward-auth/issues/11

in summary, it seemed that the redirect uri (X-Forwarded-Uri) was not being set because traefik wasn't trusting the forwarded headers, so i added this to my traefik config:

```
 - --entryPoints.web.forwardedHeaders.insecure #adding to get around redirect loop, will narrow later
- --entryPoints.websecure.forwardedHeaders.insecure #adding to get around redirect loop, will narrow later
```

This is the traefik docs on trusting the forward headers: 
https://docs.traefik.io/routing/entrypoints/#forwarded-header

if this works for you, then take off the insecure and add the domains that traefik should trust.

**Comment by EmiPhil** [comment]
Adding the insecure flags doesn't seem to fix the issue for me. Could you post the rest of your config?

**Comment by javaadpatel** [comment]
sure no problem, the Traefik setup can be found at this repository: 
https://github.com/javaadpatel/Medium_Deploying-Traefik-to-AKS/tree/master

the only difference there being that I added the 
```
- --entryPoints.web.forwardedHeaders.insecure #adding to get around redirect loop, will narrow later
- --entryPoints.websecure.forwardedHeaders.insecure #adding to get around redirect loop, will narrow later
```

and then this is my Traefik dashboard rule, to get back to the traefik dashboard as I was trying to secure the dashboard:
```
apiVersion: traefik.containo.us/v1alpha1
kind: IngressRoute
metadata:
  name: traefik-dashboard
spec:
  entryPoints:
    - web
    - websecure
  routes:
    - match: Host(`traefik.gitsensei.dev`)
      kind: Rule
      services:
        - name: api@internal
          kind: TraefikService
      middlewares:
        - name: auth
  tls:
    certResolver: le
    domains:
      - main: gitsensei.dev
        sans:
          - "*.gitsensei.dev"

---
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: auth
spec:
  forwardAuth:
    address: https://traefikauth.gitsensei.dev/_oauth
    trustForwardHeader: true
    tls:
      insecureSkipVerify: true
    authResponseHeaders:
      - X-Forwarded-User
```

and finally this is the traefik-forward-auth deployment:
```
kind: Deployment
apiVersion: extensions/v1beta1
metadata:
  name: traefikauth
  namespace: default
  labels:
    app: containous
    name: traefikauth
spec:
  replicas: 1
  selector:
    matchLabels:
      app: containous
      task: traefikauth
  template:
    metadata:
      labels:
        app: containous
        task: traefikauth
    spec:
      containers:
        - name: traefikauth
          image: thomseddon/traefik-forward-auth:2
          ports:
            - containerPort: 4181 #80
              protocol: TCP
          env:
            - name: PROVIDERS_GOOGLE_CLIENT_ID
              value: "YOUR-ID-HERE"
            - name: PROVIDERS_GOOGLE_CLIENT_SECRET
              value: "YOUR-SECRET-HERE"
            - name: SECRET
              value: "YOUR-BASE64-ENCODED-SECRET-FOR-SIGNING-JWT"
            - name: LOG_LEVEL
              value: trace
---
apiVersion: v1
kind: Service
metadata:
  name: traefikauth
  namespace: default

spec:
  ports:
    - name: http
      protocol: TCP
      port: 80
      targetPort: 4181 #adding a target port, made the requests get to container
  selector:
    app: containous
    task: traefikauth
---
apiVersion: traefik.containo.us/v1alpha1
kind: IngressRoute
metadata:
  name: traefikauth-route
  namespace: default

spec:
  entryPoints:
    - web
    - websecure

  routes:
    - match: Host(`traefikauth.gitsensei.dev`)
      kind: Rule
      services:
        - name: traefikauth
          port: 80
  tls:
    certResolver: le
    domains:
      - main: gitsensei.dev
        sans:
          - "*.gitsensei.dev"
```

apologies if its a bit rough, just got it working.

**Comment by EmiPhil** [comment]
Other than the insecure flags, your config looks similar to mine. another big difference is that I am trying to get it to work for many sub-domains which I think may be where the problem is happening.

I wrote a simple node server to attempt multi sub-domain authentication behind a traefik forward-auth middleware and got it to work _without editing any of my traefik configs._ (Host, cookie domain, and insecure cookie settings were all set to their equivalents).

So I am thinking that for some reason this module is not setting or reading the cookies properly when using sub-domains in Traefik 2+.

**Comment by samhuss** [comment]
Hi,
I'm just posting what worked with me for reference after searching for a fix for 2 days, 
I came around your issue while searching for a fix but nothing changed after applying all the comments you mentioned. My setup is Traefik + Keycloak on k3s.
 
This [blog](https://itnext.io/how-to-implement-a-sso-middleware-for-traefik-v2-on-kubernetes-dcd9d45cc875) saved me after applying everything as is then started to tune it to my setup.

I first thought the problem was in `thomseddon/traefik-forward-auth:2` and Keycloak but seems the forward-auth middleware needed special setup to bypass the loop

**Here is what I did to let it work:**
1. I started first by applying everything as is, copy/paste, only customized example.com to my domain and all worked fine with google auth.
2. Then changed the secrets and provider to be `oidc`  and let the configuration point to keycloak auth url and it also worked after commenting the `WHITELIST and DOMAIN` variables to pass the email validation and it also worked
What I was doing wrong:
- In the middleware, I was pointing to the public url with `https` instead of the internal service of `traefik-forward-auth`, once I changed the address to `address: http://traefik-forward-auth.traefik` the request started to flow from keycloak to the middleware but the callback was blocked by the middleware which was sending it back to be authenticated again
- I had the tls value in the middleware configuration to do insecure tls which was blocking the authentication callback as the callback endpoint was not using https while adding the tls  attribute seemed to activate the https scheme in communication. The callback passed after commenting this value
```
tls: 
  insecureSkipVerify: true
```  
- Added the value: `middlewares`  to the IngressRoute of `traefik-forward-auth` that expose the auth callback endpoint to the 3rd party OAuth server. Note this is the same `match` value that was used in the env variable `AUTH_HOST` of  `thomseddon/traefik-forward-auth:2` 
In case of multiple domains, this AUTH_HOST value will be added to google and keycloak to validate multiple domains. 

```
apiVersion: traefik.containo.us/v1alpha1
kind: IngressRoute
metadata:
  name: traefik-forward-auth
  namespace: traefik
spec:
  entryPoints:
    - websecure
  routes:
  - match: HOST(\`auth.example.com\`) 
    kind: Rule
    services:
    - name:  traefik-forward-auth
      port: 80
      namespace: traefik
    middlewares:
      - name: sso
  tls: 
    certResolver: myresolver
    domains:
    - main: "*.example.com
```
I hope that was useful to someone as it was for me

**Comment by UbiquitousBear** [comment]
@samhuss would you mind sharing all of your configs?

**Comment by thomseddon** [comment]
Keeping 👀 on this thread and will be looking to improve the documentation for k8s (which I use) and Traefik 2 (which I will start testing with to build docs) 👍

**Comment by SuperSandro2000** [comment]
@thomseddon feel free to contact me if you have big troubles with traefik 2 like me.

**Comment by simt2** [comment]
I've been debugging this for weeks on and off and just found the cause. The callback was actually never handled by the callback handler but by the default authentication handler. The following changes work around the rule-based routing logic via imported traefik stuff via simple `http.HandleFunc` and got my perfectly valid deployment working without any changes:
```diff
diff --git a/cmd/main.go b/cmd/main.go
index 512d70a..2e6a8c8 100644
--- a/cmd/main.go
+++ b/cmd/main.go
@@ -21,7 +21,8 @@ func main() {
        server := internal.NewServer()

        // Attach router to default server
-       http.HandleFunc("/", server.RootHandler)
+       http.HandleFunc("/", server.AuthHandler(config.DefaultProvider, "default"))
+       http.HandleFunc(config.Path, server.AuthCallbackHandler())

        // Start
        log.Debugf("Starting with options: %s", config)
diff --git a/internal/server.go b/internal/server.go
index 9027a47..67163ef 100644
--- a/internal/server.go
+++ b/internal/server.go
@@ -15,7 +15,6 @@ type Server struct {

 func NewServer() *Server {
        s := &Server{}
-       s.buildRoutes()
        return s
 }

@@ -70,6 +69,10 @@ func (s *Server) AuthHandler(providerName, rule string) http.HandlerFunc {
        p, _ := config.GetConfiguredProvider(providerName)

        return func(w http.ResponseWriter, r *http.Request) {
+               r.Method = r.Header.Get("X-Forwarded-Method")
+               r.Host = r.Header.Get("X-Forwarded-Host")
+               r.URL, _ = url.Parse(r.Header.Get("X-Forwarded-Uri"))
+
                // Logging setup
                logger := s.logger(r, rule, "Authenticating request")

```

This is a bug I' not eager to hunt down.

**Comment by thomseddon** [comment]
@simt2 could you post your rules config?

As mentioned above, there is a known bug identified in #101 whereby certain rules can break the authentication flow.
This is the wrong behaviour, but I think changing this would be backwards incompatible as it is conceivable that some users may have done this deliberately. I will add a note in the README to warn of this when using rules!

**Comment by simt2** [comment]
@thomseddon  I did not configure any rules. My traefik configuration:
traefik.yml
```
accessLog:
  format: json
entryPoints:
  traefik:
    address: ":8080"
  web:
    address: ":80"
  websecure:
    address: ":443"
    http:
      tls: {}
defaultEntrypoints:
- websecure
api:
  dashboard: true
  # Protected via google auth from outside.
  insecure: true
ping: true
log:
  level: info
  format: json
providers:
  kubernetesIngress:
    ingressEndpoint:
      publishedService: $(TRAEFIK_EXTERNAL_SERVICE_NAMESPACE)/$(TRAEFIK_EXTERNAL_SERVICE)
  file:
    filename: /config/traefik.dynamic.yml
```

traefik.dynamic.yml
```
http:
  routers:
    http_catchall:
      priority: 100
      entryPoints:
      - web
      middlewares:
      - redirecthttps
      rule: "HostRegexp(`{any:.+}`)"
      service: noop@internal
  middlewares:
    google-auth:
      forwardAuth:
        address: "http://traefik-forward-auth:4181/"
        authResponseHeaders:
        - "X-Forwarded-User"
        trustForwardHeader: true
    redirecthttps:
      redirectScheme:
        scheme: https
        permanent: true
```

deployment:
```
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: traefik
  labels:
    component: traefik
spec:
  replicas: 1
  selector:
    matchLabels:
      component: traefik
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
  template:
    metadata:
      labels:
        component: traefik
    spec:
      serviceAccountName: traefik
      terminationGracePeriodSeconds: 60
      containers:
      - image: traefik:v2.2.1
        name: traefik
        resources:
        readinessProbe:
          httpGet:
            path: /ping
            port: 8080
          failureThreshold: 1
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 2
        livenessProbe:
          httpGet:
            path: /ping
            port: 8080
          failureThreshold: 3
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 2
        ports:
        - name: traefik
          containerPort: 8080
          protocol: TCP
        - name: http
          containerPort: 80
          protocol: TCP
        - name: https
          containerPort: 443
          protocol: TCP
        args:
        - "--configfile=/config/traefik.yml"
        env:
        - name: TRAEFIK_PROVIDERS_KUBERNETESINGRESS_INGRESSENDPOINT_PUBLISHEDSERVICE
          value: $(TRAEFIK_EXTERNAL_SERVICE_NAMESPACE)/$(TRAEFIK_EXTERNAL_SERVICE)
        volumeMounts:
        - mountPath: /config
          name: traefik
      volumes:
      - name: traefik
        configMap:
          name: traefik
```
Dashboard ingress protected by google auth:
```
---
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: traefik
  labels:
    component: traefik
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt-prod
    traefik.ingress.kubernetes.io/router.middlewares: google-auth@file
spec:
  rules:
  - host: traefik.internal.example.com
    http:
      paths:
      - backend:
          serviceName: traefik
          servicePort: traefik
  tls:
  - hosts:
    - traefik.internal.example.com
    secretName: traefik-tls
```

auth.internal.example.com ingress not protected by google auth:
```
---
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: traefik-forward-auth
  labels:
    component: traefik-forward-auth
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt-prod
spec:
  rules:
  - host: auth.internal.example.com
    http:
      paths:
      - backend:
          serviceName: traefik-forward-auth
          servicePort: forward-auth
  tls:
  - hosts:
    - auth.internal.example.com
    secretName: traefik-forward-auth-tls
```

**Comment by simt2** [comment]
Not using conventional mode instead auf auth host worked fine.

**Comment by thomseddon** [comment]
@simt2 is that now actually using the auth domain? If you could also post your traefik-forward-auth deployment+config that would be helpful.

For everyone else in the thread - I have, at long last, just pushed some examples of deploying with traefik forward auth, it's not particularly complicated but it does require a good understanding of how the authenticated/unauthenticated requests flow internally.

I have checked the examples against configs posted here, and it mostly matches up. Full examples are here: https://github.com/thomseddon/traefik-forward-auth/tree/master/examples/traefik-v2/kubernetes

**Comment by simt2** [comment]
My traefik-forward-auth deployment is as simple as:
```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: traefik-forward-auth
  labels:
    component: traefik-forward-auth
spec:
  replicas: 1
  selector:
    matchLabels:
      component: traefik-forward-auth
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
  template:
    metadata:
      labels:
        component: traefik-forward-auth
    spec:
      terminationGracePeriodSeconds: 30
      containers:
      - image: thomseddon/traefik-forward-auth:2
        imagePullPolicy: Always
        name: forward-auth
        resources:
        ports:
        - name: forward-auth
          containerPort: 4181
          protocol: TCP
        args:
        - "--domain=example.com"
        - "--cookie-domain=example.com"
        - "--log-level=debug"
        - "--auth-host=auth.internal.example.com"
        envFrom:
        - secretRef:
            name: traefik-forward-auth-credentials
```

**Comment by simt2** [comment]
My understanding is that without any rule configuration, every path/request should be redirected, except `/_oauth` which should be handled by the function returned by `server.AuthCallbackHandler()`. Please correct me if I'm wrong, but this is not the case:


```shell
# Running on latest master
$ git rev-parse HEAD                                                                                                                                             
1ac0ca97321245406ed6c6593000ad5c071b3e0c
$ go run cmd/main.go --secret=bla --providers.google.client-id=1 --providers.google.client-secret=2 --log-level=debug
time="2020-05-08T11:00:04+02:00" level=debug msg="Starting with options: {\"LogLevel\":\"debug\",\"LogFormat\":\"text\",\"AuthHost\":\"\",\"CookieDomains\":null,\"InsecureCookie\":false,\"CookieName\":\"_forward_auth\",\"CSRFCookieName\":\"_forward_auth_csrf\",\"DefaultAction\":\"auth\",\"DefaultProvider\":\"google\",\"Domains\":null,\"LifetimeString\":43200,\"Path\":\"/_oauth\",\"Whitelist\":null,\"Providers\":{\"Google\":{\"ClientID\":\"1\",\"Scope\":\"https://www.googleapis.com/auth/userinfo.profile https://www.googleapis.com/auth/userinfo.email\",\"Prompt\":\"\",\"LoginURL\":{\"Scheme\":\"https\",\"Opaque\":\"\",\"User\":null,\"Host\":\"accounts.google.com\",\"Path\":\"/o/oauth2/auth\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\"},\"TokenURL\":{\"Scheme\":\"https\",\"Opaque\":\"\",\"User\":null,\"Host\":\"www.googleapis.com\",\"Path\":\"/oauth2/v3/token\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\"},\"UserURL\":{\"Scheme\":\"https\",\"Opaque\":\"\",\"User\":null,\"Host\":\"www.googleapis.com\",\"Path\":\"/oauth2/v2/userinfo\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\"}},\"OIDC\":{\"Config\":null,\"IssuerURL\":\"\",\"ClientID\":\"\"}},\"Rules\":{},\"Lifetime\":43200000000000,\"CookieDomainsLegacy\":null,\"CookieSecureLegacy\":\"\",\"ClientIdLegacy\":\"\",\"PromptLegacy\":\"\"}"
time="2020-05-08T11:00:04+02:00" level=info msg="Listening on :4181"
time="2020-05-08T11:00:26+02:00" level=debug msg="Authenticating request" headers="map[Accept:[*/*] User-Agent:[curl/7.64.1]]" rule=default source_ip=
time="2020-05-08T11:00:26+02:00" level=debug msg="Set CSRF cookie and redirecting to google login" source_ip=
time="2020-05-08T11:00:26+02:00" level=debug msg=Done source_ip=
time="2020-05-08T11:00:31+02:00" level=debug msg="Authenticating request" headers="map[Accept:[*/*] User-Agent:[curl/7.64.1]]" rule=default source_ip=
time="2020-05-08T11:00:31+02:00" level=debug msg="Set CSRF cookie and redirecting to google login" source_ip=
time="2020-05-08T11:00:31+02:00" level=debug msg=Done source_ip=
```

```shell
$ curl http://localhost:4181/should_be_redirected
$ curl http://localhost:4181/_oauth 
```

**Comment by thomseddon** [comment]
@simt2 the original request that's being proxied by traefik is read from the `X-Forwarded-Method`, `X-Forwarded-Host` and `X-Forwarded-Uri` headers, so to repeat your example:

```
$ curl -I -H 'X-Forwarded-Uri: /should_be_redirected' 127.0.0.1:4181
HTTP/1.1 307 Temporary Redirect
Location: https://accounts.google.com/o/oauth2/auth?client_id=1&redirect_uri=%3A%2F%2F%2F_oauth&response_type=code&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email&state=f59cebc5c22e1d33b0e94adf913624f2%3Agoogle%3A%3A%2F%2F%2Fshould_be_redirected
Set-Cookie: _forward_auth_csrf=f59cebc5c22e1d33b0e94adf913624f2; Path=/; Expires=Sat, 09 May 2020 07:05:57 GMT; HttpOnly; Secure
Date: Fri, 08 May 2020 19:05:57 GMT
Content-Length: 0

$ curl -I -H 'X-Forwarded-Uri: /_oauth' 127.0.0.1:4181
HTTP/1.1 401 Unauthorized
Content-Type: text/plain; charset=utf-8
X-Content-Type-Options: nosniff
Date: Fri, 08 May 2020 19:05:41 GMT
Content-Length: 15
```

We also have a number of tests in place to ensure valid/invalid requests are permitted/denied as expected.

I'm still not 100% sure but I believe your configuration may not have worked as you have included a trailing slash in your middleware definition, could you try without? (example middleware now available here: https://github.com/thomseddon/traefik-forward-auth/blob/master/examples/traefik-v2/kubernetes/advanced-separate-pod/traefik-forward-auth/middleware.yaml)

**Comment by thomseddon** [comment]
I believe the issue for most people in this thread will have been caused by configuration issues due to there not being and good docs on how to configure tfa+traefik v2+k8s. Hopefully this is now fixed, with a few complete examples here:  https://github.com/thomseddon/traefik-forward-auth/blob/master/examples/traefik-v2/kubernetes

I'm still not 100% sure what was causing your issue @simt2, the examples above have been tested and all seem to work and there are a number of others who have been able to get this up and working in the environment. Please can I suggest that if you're unable to work through your issues based on the new documentation, please open a new issue with a complete minimal example of all the resources you're deploying to your cluster and I will attempt to recreate locally.

**Comment by simt2** [comment]
@thomseddon Thanks for the extensive examples! 

I did try without the slash. The requests outgoing from my traefik v2 instance to traefik-forward-auth don't contain a `X-Forwarded-Uri` header at all, leading to the requests URL to be overwritten with an empty string. I'll look into that.

**Comment by DaazKu** [comment]
If you ever a redirect problem too here is the solution: https://github.com/thomseddon/traefik-forward-auth/issues/198#issuecomment-902513612

`:latest` != `:2.2.0` or `:2`
![image](https://user-images.githubusercontent.com/2412909/199360905-e3667832-e197-49b8-9daa-70ff32591e9b.png)

2.2.0 has a bug and will redirect you over and over.
